### PR TITLE
ENH: add nearest-up rounding to scipy.interpolate.interp1d

### LINKED ISSUE #12749: ENH: interp1d Matlab compatibility
**Is your feature request related to a problem? Please describe.**
Yes. I want to rewrite Matlab code in Python. Unfortunately the interp1d with 'nearest' is not fully compatible with Matlab interp1. Matlab rounds up and Scipy rounds down. If one wants to interpolate 1.5 between 1 and 2, Matlab gives you the value at 2 and Scipy gives you the value at 1. That's really a mess, if one expects same results.

**Describe the solution you'd like**
An additional parameter left/right or nearest_up/nearest_down

### PR DESCRIPTION
**Pull request by skresearcher**
#### Reference issue
Closes #12749.

#### What does this implement/fix?
Add nearest-up rounding as an option to scipy.interpolate.interp1d.

### CODE REVIEW

**Comment by skresearcher** [comment]
The Azure Pipelines for 32-bit Python 3.6 keeps bringing this error:

```
Failures
 - mingw (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\mingw\tools\chocolateyinstall.ps1'.
 See log for details.
##[error]PowerShell exited with code '1'.
```

I am not sure whether this is because of my changes or it is unrelated.

**Comment by skresearcher** [comment]
@ev-br if you have time, can you please review this PR?

**Review by ev-br — CHANGES_REQUESTED** [review]

File: scipy/interpolate/tests/test_interpolate.py [hunk 1]
```diff
@@ -322,6 +322,25 @@ def test_nearest(self):
                     bounds_error=True)
         assert_raises(ValueError, interp1d, self.x10, self.y10, **opts)
 
+    def test_nearest_up(self):
+        # Check the actual implementation of nearest-neighbour interpolation.
+        interp10 = interp1d(self.x10, self.y10, kind='nearest-up')
+        assert_array_almost_equal(interp10(self.x10), self.y10)
+        assert_array_almost_equal(interp10(1.2), np.array(1.))
+        assert_array_almost_equal(interp10([2.4, 5.6, 6.0]),
+                                np.array([2., 6., 6.]),)
```
[the remark this answers is not in the document]
Comment (skresearcher) [reply scipy/interpolate/tests/test_interpolate.py]: Fixed.

**Comment by skresearcher** [comment]
@ev-br I've made the changes which you requested. It would be great to get this included in the next release (1.6.0).

**Comment by skresearcher** [comment]
@ev-br @tylerjereddy following up on this PR review. This is a small change to the interp1d function which I'm hoping can be pushed with the next release of SciPy.

**Comment by tylerjereddy** [comment]
I'll add the milestone tentatively; we'll see what Evgeni says

File: scipy/interpolate/interpolate.py [hunk 2]
```diff
@@ -343,13 +343,12 @@ class interp1d(_Interpolator1D):
...
-        order; 'previous' and 'next' simply return the previous or next value
-        of the point) or as an integer specifying the order of the spline
-        interpolator to use.
-        Default is 'linear'.
+        ('linear', 'nearest', 'nearest-up', 'zero', 'slinear', 'quadratic',
+        'cubic', 'previous', 'next', where 'zero', 'slinear', 'quadratic'
+        and 'cubic' refer to a spline interpolation of zeroth, first,
+        second or third order; 'previous' and 'next' simply return the
+        previous or next value of the point) or as an integer specifying
+        the order of the spline interpolator to use. Default is 'linear'.
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:351]: This part needs a one-sentence explanation along the lines of ``"nearest-up" differs from "nearest" in that it ...'' or some such.

File: scipy/interpolate/tests/test_interpolate.py [hunk 3]
```diff
@@ -322,11 +324,32 @@ def test_nearest(self):
                     bounds_error=True)
         assert_raises(ValueError, interp1d, self.x10, self.y10, **opts)
 
+    def test_nearest_up(self):
+        # Check the actual implementation of nearest-neighbour interpolation.
```
Comment (ev-br) [inline scipy/interpolate/tests/test_interpolate.py:328]: Can/should expand this comment, and again stress the difference between nearest and nearest-up

**Review by ev-br — APPROVED** [review]
So, the only difference in nearest and nearest-up is how it deals with half-integer input, correct? This needs to be explained in a couple of places, which I marked inline. These tweaks can be made in a follow-up PR though.

File: scipy/interpolate/interpolate.py (hunk 2 above)
Comment (skresearcher) [reply scipy/interpolate/interpolate.py]: Fixed.

File: scipy/interpolate/tests/test_interpolate.py (hunk 3 above)
Comment (skresearcher) [reply scipy/interpolate/tests/test_interpolate.py:329]: Fixed.

**Comment by ev-br** [comment]
Test failures/stalls are unrelated, let's get this in. Thanks @skresearcher , Tyler.

**Comment by ev-br** [comment]
A short release note under `scipy.interpolate improvements` would be useful to add here:  
https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0

**Pull request merged**